### PR TITLE
AMBER Energy Functions

### PR DESCRIPTION
**Pull request by alecbetancourt**
This draft PR aims to add the AMBER potential energy functions along with some additional MM simulation related features to JAX-MD. Our lab maintains the ReaxFF pull request that was recently merged along with the JAX-ReaxFF optimization suite. We've been internally using this AMBER code for torsion parameter optimization and some hybrid reactive/MM simulations. We are also looking to add missing AMBER features as necessary to make the code more usable for the general public.

- mm_forcefields/amber/energy.py
  - AMBER/CHARMM energy terms. This should cover the basic set of terms used for FF19SB, GAFF/GAFF2, and CHARMM36 along with any other derivative forcefields. In addition to the standard bond/angle/torsion/coulomb/LJ terms, analytic 12-6 LJ dispersion correction, CMAPs, NBFIX exceptions, Improper terms, and Urey-Bradley terms are also supported.

- mm_forcefields/amber/constraints.py
  - CCMA/SETTLE to handle general holonomic distance constraints and rigid water constraints. I can also implement SHAKE, RATTLE, or any other desired algorithms for this, but CCMA is generally one of the best performing to my knowledge.

- mm_forcefields/nonbonded/electrostatics.py
  - I based my code on the original PME implementation by Sam. The implementation in this version of electrostatics module was quite different and missing some features. In particular, the lack of B-spline interpolation, fractional coordinate/triclinic box support, use of N^2 masks and some other sharp edges made it unsuitable for my application. I've ported over the original PME implementation and added additional support for order 5 spline interpolation, perturbations for NPT support, as well as a few other small changes. In addition, for the OPLS code, exclusions and 1-4 exceptions are handled explicitly with post hoc masking. I've tried to avoid this for the most part and treat both as general exceptions that are masked out of the neighbor list and handled based on their exception parameters, including for PME correction bookkeeping. I think this is the most generally applicable approach to this problem if there is the possibility of having NBFIX style exclusions or some mix of zero/non-zero LJ/Coulomb. I've mirrored the approach used in OpenMM and internally in AMBER, but let me know if there are other idiomatic ways of doing this. I also generally think the structure of this class isn't very amenable to the functional programming style used for most of the other energy functions (i.e. stateless energy function generator + dataclass).

- mm_forcefields/base.py & mm_forcefields/oplsaa/params.py
  - I tried to reuse the interface from OPLS as much as possible. The potentials are similar enough that I think most of the code should be based on a common set of data structures at some point. I think most of the classes in here should eventually be converted to dataclass instances from dataclasses.py.

- mm_forcefields/io/openmm.py
  - I don't think writing one off input parsers is generally a good idea. Systems can be parameterized in a variety of ways for a given forcefield and can include a number of different file formats with legacy compatibility issues. ParmEd and OpenFF/OpenMM have the most complete and canonical parsers that I'm aware of for AMBER/CHARMM like systems. In addition, there is a template generator through OpenFF that allows the generation of small molecule parameters via a variety of methods (GAFF/Smirnoff/Espaloma). I've specifically written a parser that takes an OpenMM system as an input and converts it into a JAX compatible dataclass.

- partition.py
  - In the event a more complex energy function supports both periodic and non periodic systems, I think it's most natural to have an idempotent interface when using smap. If energy terms are implemented in terms of smap, you can run into NOBC with/without a cutoff, PBC with a cutoff, and PBC with a unit cell that is small enough that a fallback N^2 list with shifts for multiple images will be needed. I think it makes sense from an interface perspective to be able to treat all of these situations with partition.neighbor_list and then map neighbor-like pair potentials through smap.pair_neighbor_list. As a matter of uniformity, I think it would help for every smap function to eventually accept the same types of mapping schemes.

- smap.py
  - Instead of doing ad-hoc masking, it seems like the smap interface is a better way to deal with pair potentials in general. It seemed like a natural extension to add support for 3 and 4 body interactions. It may even be a good idea to modify these to accept pairs of pairs via broadcasting or passing a trial shape (e.g. bond correlation functions or Phi/Psi pairs for CMAP evaluation). I've also added the binary function mapping from smap.pair_neighbor_list to smap.bond.

- tests/amber_test.py
  - The current tests should have fairly good coverage and energy agreement for most systems is at least within 1e-5 relative tolerance. Although the PME implementation is generally correct, it may still be possible to get even better agreement. More tests for longer trajectories would also help to validate longer term properties of interest. The current suite takes about 10 minutes on a V100, so performance is a concern.

- data/amber_data
  - This directory contains the test systems. There is a small molecule parameterized with GAFF (Sustiva), RAMP1 protein in a variety of water boxes and periodic cells to test virtual site and constraint handling, as well as some of the systems (Both PME/GB) from the Amber20 GPU benchmark suite.

Some tasks that need to be completed before merging include:
- [ ] Linting and test coverage checks
- [ ] A review of the interface changes for the MM interface and ideally more standardization
- [ ] Add type hints and documentation for functions and class structure
- [ ] Add more extensive automated testing on longer simulations
- [ ] Providing Jupyter notebooks for simulations using the energy functions as well as a more advanced example for something like generating torsional scans
- [ ] Performance profiling to improve comparisons to OpenMM (currently slower by ~7x)

In addition, some nice features to implement as needed could include:
- [ ] Better parser for writing out modified forcefields back to valid prmtop, ParmEd structure, or other format
- [ ] Implicit solvent and periodic cutoff functions if needed
- [ ] Single/double/mixed precision modes to better match other MD libraries and improve performance

Any feedback or feature suggestions are appreciated and changes will be posted as the PR is cleaned up for submission.

### CODE REVIEW

**Comment by abhijeetgangan** [comment]
Hi @beitom @alecbetancourt ,

I can help with merging the PR. If anything is pending I can help with it.

**Comment by ekindogus** [comment]
jax-md isn't dead! There are several large scale projects using jax-md, so it should get less dead moving forward. 

Please let us know if there is anything we can help with.

**Comment by alecbetancourt** [comment]
Hi @beitom @ekindogus ,

I've also been building out this PR a little more lately to fully replicate the behavior of Sander when run as a standalone tool. Getting this PR merged had taken a bit of a backseat with the slow down of activity on this repository and our focus on our parameter optimization tool, but I'm glad to know that things are still active. If it would be easier to discuss this with a meeting, I'd be happy to sit down and go over this in more detail. Feel free to reach out to me at my personal email (alecbetancourt@gmail.com). I saw your comment on the OPLS PR and I'm interested in figuring out the most ergonomic way to package all of this going forward. The code as it stands is capable of running normal simulations, PME, and some basic restraints, but I think there are likely some changes that should be made if the goal is to target a more general environment for MM force fields.

**Comment by alecbetancourt** [comment]
Hi all, apologies for the lack of responsiveness the last few weeks. I was busy finishing my internship in Japan and had very little bandwidth for anything else. I'm back in the US now and am going to be looking into this during the week. I need to do some substantial restructuring of the code to fit the new interface and clean up some rough edges for features that were partially implemented or implemented but not well tested.

**Comment by abhijeetgangan** [comment]
Hi @alecbetancourt , thanks for the update! I will review in detail over the weekend. Quick comments:
- You need to lint the code.
- Are the data files required? I guess one is fine for testing.

**Comment by abhijeetgangan** [comment]
@alecbetancourt I am trying to run the perf script but it is failing (see error in CI). Could you fix it?

**Comment by alecbetancourt** [comment]
@abhijeetgangan 
> @alecbetancourt I am trying to run the perf script but it is failing (see error in CI). Could you fix it?

My apologies for completely missing this reply. I think the extra dependencies causing the build failures go into pyproject.toml in this case, but let me know if I'm wrong. These additions require a few packages (e.g. OpenMM/ParmEd for parsing input files). I set up separate test jobs for the AMBER related additions and am running them on my fork right now.

We had discovered some issues regarding numerical agreement with other MD packages that we weren't entirely happy with. I've done some more work here as well as fixing most of the coverage lapses and tightening most of the test tolerances. I've also added in an example for constructing alchemically modified potentials to do hydration free energy simulations (hopefully this can eventually be extended to binding free energy and other applications). Things are at a good point now in terms of numerical agreement for a variety of systems with different solvent models, AMBER/CHARMM forcefields, dispersion corrections, virtual sites, constraints, etc.

I need to update some of these details in the main message for the pull request and clean up the test files/move them to another repository but I think everything should be fairly close to a suitable state for a first release of this code. Let me know if anything else is missing!

**Comment by abhijeetgangan** [comment]
Thanks @alecbetancourt I will take a look into it and sorry for the late reply I was planning to get the UMA PR in. I think it is done.

> My apologies for completely missing this reply. I think the extra dependencies causing the build failures go into pyproject.toml in this case, but let me know if I'm wrong. These additions require a few packages (e.g. OpenMM/ParmEd for parsing input files). I set up separate test jobs for the AMBER related additions and am running them on my fork right now.

No worries. I will take a look at the deps issues. If we want to merge it it would and if they have heavy deps then we can make a dep group for it.

> We had discovered some issues regarding numerical agreement with other MD packages that we weren't entirely happy with. I've done some more work here as well as fixing most of the coverage lapses and tightening most of the test tolerances. I've also added in an example for constructing alchemically modified potentials to do hydration free energy simulations (hopefully this can eventually be extended to binding free energy and other applications). Things are at a good point now in terms of numerical agreement for a variety of systems with different solvent models, AMBER/CHARMM forcefields, dispersion corrections, virtual sites, constraints, etc.

Thanks for fixing it. Could you tell me which cases had divergence from other MD packages? RE the latter: That sounds good!

> I need to update some of these details in the main message for the pull request and clean up the test files/move them to another repository but I think everything should be fairly close to a suitable state for a first release of this code. Let me know if anything else is missing!

Thanks a lot! I think there is one thing that would be nice is to put all the data files somewhere. I am not sure what is the place, it makes the PR look big.

Also, if you don't mind can I push some changes if needed and also let codex review bot run?

**Comment by alecbetancourt** [comment]
@abhijeetgangan 

No worries at all! 

Regarding your second comment, we were seeing good agreement for simple, non-periodic systems, but with more complex systems, there were a handful of issues with the energy/force tests. These were mostly related to virtual-site handling (and the associated force redistribution from the reduced DoFs), PME interpolation order, and switching/dispersion correction (this can vary from system to system depending on which forces these are turned on for, softcore decoupling along a lambda schedule, etc.). We now see good energy/force agreement across a variety of systems with OpenMM as well as a good match to GROMACS for the softcore/free energy related tests.

I went ahead and compressed the data files into a tarball that is decompressed by the test runner. This substantially reduced the number of changed files and it looks like on my fork, all of the test actions still pass without any issues. Feel free to add any changes you see fit and run any other reviews that may be helpful. Let me know if there are any issues or interface changes that need to be made and I'm happy to work on ironing them out!

**Comment by abhijeetgangan** [comment]
@alecbetancourt Thanks for the info! If you can give me the reproducers if any of the errors still persist then I can take a look.

Regarding the code changes. I think everything looks good. I think the changes to smap and partition should be a separate PR. Other things can be in this PR.

**Comment by abhijeetgangan** [comment]
@codex review

File: jax_md/partition.py [hunk 1]
```diff
@@ -950,11 +967,221 @@ def prune_neighbor_list_sparse(
...
+      receiver = jnp.reshape(neigh, (BLOCK, M))
+
+      pos_sender = position[sender_safe]  # (B, dim)
+      receiver_safe = jnp.minimum(receiver, N_i32 - i32(1))
+      pos_receiver = position[receiver_safe]  # (B, M, dim)
+      dr2 = d(pos_sender, pos_receiver)  # (B, M)
+
+      mask = valid_sender[:, None] & (receiver < N_i32) & (dr2 < cutoff_sq)
+      if not mask_self:
+        mask = mask & (receiver != sender[:, None])
```
[the remark this answers is not in the document]
Comment (alecbetancourt) [reply jax_md/partition.py]: Not a concern as the partition code in this PR will be reverted pending the other open PR that is fixing the memory issues with the current implementation.

File: jax_md/mm_forcefields/amber/energy.py [hunk 2]
```diff
@@ -0,0 +1,1436 @@
...
+  elif nb_options.r_cut is not None:
+    vdw_cutoff_fn = partial(hard_cutoff, r_cut=nb_options.r_cut)
+    ele_cutoff_fn = partial(hard_cutoff, r_cut=nb_options.r_cut)
+  else:
+    vdw_cutoff_fn = lambda fn: fn
+    ele_cutoff_fn = lambda fn: fn
+
+  # Create fused LJ + coulomb function to avoid 2 passes over smap
+  # NOTE probably not the best way to flag if the ab form is used
+  use_ab = len(topology.nbfix_atom_type) != 0
```
[the remark this answers is not in the document]
Comment (alecbetancourt) [reply jax_md/mm_forcefields/amber/energy.py:468]: In this case, the OpenMM conversion scripts never set this to ```None```. There are a few things like this that I'm not particularly fond of, but this was done to avoid causing issues with the OPLS implementation that uses the same underlying data structures. This should be fine for now pending a cleanup of the MM interface.

File: jax_md/mm_forcefields/base.py [hunk 3]
```diff
@@ -94,3 +125,168 @@ class NonbondedOptions(NamedTuple):
...
   scale_14_coul: float = 0.5
+  nb_format: NeighborListFormat = (
+    NeighborListFormat.Dense
+  )  # TODO make all optional
+  use_pbc: bool = True
+  use_periodic_general: bool = False
+  fractional_coordinates: bool = False
+  wrapped_space: bool = True
+  disp_coef: float = 0.0
+  r_switch: float = 0.0
```
[the remark this answers is not in the document]
Comment (alecbetancourt) [reply jax_md/mm_forcefields/base.py:136]: This is a similar case to the suggestion for energy.py. The default values are set based on what was already in base.py for the OPLS implementation but the system conversion functions for the AMBER code set ```r_switch``` correctly depending on the nonbonded/custom forces in the OpenMM system. This should also be revisited in the future, but shouldn't be an issue for now.

File: jax_md/mm_forcefields/amber/energy.py [hunk 4]
```diff
@@ -0,0 +1,1446 @@
...
+
+    coul_dir_pot = jnp.sum(coul_dir_edge)
+
+    if use_softcore_vdw and use_linear_coul:
+      assert pair_coul_1r_edge_fn is not None
+      sc_sc_mask, _, _ = _sc_edge_masks(nbr_list)
+      sc_sc_coul_1r_edge = pair_coul_1r_edge_fn(
+        positions,
+        nbr_list,
+        charge_sq=params.nonbonded.charges,
```
[the remark this answers is not in the document]
Comment (alecbetancourt) [reply jax_md/mm_forcefields/amber/energy.py:890]: This is a good catch. Fixed.
